src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -96,12 +97,107 @@ public static void Fill(Span<byte> data)
             RandomNumberGeneratorImplementation.FillSpan(data);
         }
 
+        public static int GetInt32(int fromInclusive, int toExclusive)
+        {
+            if (fromInclusive >= toExclusive) throw new ArgumentException(SR.Argument_InvalidRandomRange);
+
+            // The total possible range is [0, 4,294,967,295).
+            long adjustedRange = (long)toExclusive - fromInclusive;
+
+            // If there is only one possible choice, nothing random will actually happen, so return
+            // the only possibility.
+            if (adjustedRange == 1)
+            {
+                return fromInclusive;
+            }
+
+            // Subtract one to account for zero being an actual possibility. We know for certain that
+            // the range fits in a UInt32 without overflowing.
+            uint maskedRange = (uint)adjustedRange - 1;
+
+            int leadingZeroCount = LeadingZeroCount(maskedRange);
+
+            // Create a mask for the bits that we care about for the range. The other bits will be
+            // masked away.
+            uint mask;
+
+            // If the LZCNT is zero create a full mask to avoid overflowing when shifting.
+            if (leadingZeroCount == 0)
+            {
+                mask = unchecked((uint)-1);
+            }
+            else
+            {
+                mask = (1U << 32 - leadingZeroCount) - 1;
+            }
+
+            // We only want to generate as many bytes as required to satisfy the mask to not apply
+            // undue pressure to the underlying random number generator.
+            int bytesRequired = (32 + 7 - leadingZeroCount) / 8;
+
+            Span<uint> valueResult = stackalloc uint[1];
+            ref uint result = ref valueResult[0];
+            Span<byte> valueResultBytes = MemoryMarshal.AsBytes(valueResult);
+            if (BitConverter.IsLittleEndian)
+            {
+                do
+                {
+                    Span<byte> randomFillSlice = valueResultBytes.Slice(0, bytesRequired);
+                    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
+                }
+                while ((result & mask) > maskedRange);
+            }
+            else
```

Comment:
I don't have any strong opinions about how to handle big endian archs here or the appropriate way to test it. You could also use the same slicing from the little endian path and use `BinaryPrimitives.ReverseEndianness`. I thought BE should avoid the penalty though.

Reply:
The actual random number generation is going to dominate the performance of this method. I doubt that you are getting any measurable speed up by micro-optimizing this for BE/LE. 

It would be better to make this simpler and smaller code, with no special casing for BE/LE, something like:

```
do
{
    Span<byte> randomFillSlice = valueResultBytes.Slice(0, bytesRequired);
    RandomNumberGeneratorImplementation.FillSpan(randomFillSlice);
    result = BinaryPrimitives.ReadInt32LittleEndian(randomFillSlice);
}
while ((result & mask) > maskedRange);
```

Reply:
Bah. That makes way more sense. Thanks!